test/cpp/util/channelz_sampler.cc

```diff
@@ -0,0 +1,476 @@
+    std::string type = "Subchannel";
+    std::string description;
+    ::google::protobuf::TextFormat::PrintToString(subchannel.data(),
+                                                  &description);
+    StoreEntityInJson(id, type, description);
+  }
+
+  // Store a server in Json
+  void StoreServerInJson(const grpc::channelz::v1::Server& server) {
+    std::string id = grpc::to_string(GetServerID(server));
+    std::string type = "Server";
+    std::string description;
+    ::google::protobuf::TextFormat::PrintToString(server.data(), &description);
+    StoreEntityInJson(id, type, description);
+  }
+
+  // Store a socket in Json
+  void StoreSocketInJson(const grpc::channelz::v1::Socket& socket) {
+    std::string id = grpc::to_string(GetSocketID(socket));
+    std::string type = "Socket";
+    std::string description;
+    ::google::protobuf::TextFormat::PrintToString(socket.data(), &description);
+    StoreEntityInJson(id, type, description);
+  }
+
+  // Store an entity in Json
+  void StoreEntityInJson(std::string& id, std::string& type,
+                         std::string& description) {
+    std::string start, finish;
+    time_t ago = now - int64_t(FLAGS_sampling_interval_seconds);
+    std::stringstream ss;
+    ss << std::put_time(std::localtime(&now), "%F %T");
+    finish = ss.str();  // example: "2019-02-01 12:12:18"
+    ss.str("");
+    ss << std::put_time(std::localtime(&ago), "%F %T");
+    start = ss.str();
+
+    grpc_core::Json obj =
+        grpc_core::Json::Object{{"Task", absl::StrFormat("%s_ID%s", type, id)},
+                                {"Start", start},
+                                {"Finish", finish},
+                                {"ID", id},
+                                {"Type", type},
+                                {"Description", description}};
+    json.mutable_array()->push_back(obj);
+  }
+
+  // Dump data in json
+  std::string DumpJson() { return json.Dump(); }
+
+  time_t now;
+  // private:
+  std::unique_ptr<grpc::channelz::v1::Channelz::Stub> channelz_stub;
+  std::vector<grpc::channelz::v1::Channel> top_channels;
+  std::vector<grpc::channelz::v1::Server> all_servers;
+  std::vector<grpc::channelz::v1::Channel> all_channels;
+  std::vector<grpc::channelz::v1::Subchannel> all_subchannels;
+  std::vector<grpc::channelz::v1::Socket> all_sockets;
+  grpc_core::Json json;
+};
+
+int main(int argc, char** argv) {
+  // make sure flags can be used
+  grpc::testing::InitTest(&argc, &argv, true);
+
+  // create a channelz client
+  ChannelzSampler channelz_sampler;
+  channelz_sampler.Setup(FLAGS_custom_credentials_type, FLAGS_server_address);
+
+  // Keep sampling based on FLAGS_sampling_times and
+  // FLAGS_sampling_interval_seconds Sample in the end of an interval
+  for (int i = 0; i < FLAGS_sampling_times; ++i) {
+    std::cout << "Wait for sampling interval "
+              << FLAGS_sampling_interval_seconds << "s..." << std::endl;
+    sleep(FLAGS_sampling_interval_seconds);
```

Comment:
I think this sleep should come after the sample, to make the "times=1" case faster

Reply:
change the default sampling times to 1 and sampling interval to 0, so we can take a snapshot without waiting.